tao_compiler/mlir/disc/tools/disc-transform/TransformOps/TransformOpsExt.cc

```diff
@@ -575,6 +577,309 @@ DiagnosedSilenceableFailure FoldProducerExtractSliceOp::applyToOne(
+
+  // Create a GenericOp to transpose `inputTensor` into `outputTensor`.
+  auto transposeOp = b.create<linalg::GenericOp>(
+      loc, resultTensorType, inputTensor, outputTensor,
+      b.getAffineMapArrayAttr(indexingMaps), b.getStrArrayAttr(iteratorTypes),
+      /*doc=*/nullptr,
+      /*library_call=*/nullptr);
+  Region& body = transposeOp.getRegion();
+  body.push_back(new Block());
+  body.front().addArguments({elementType, elementType}, {loc, loc});
+
+  // Create the body of the transpose operation.
+  OpBuilder::InsertionGuard g(b);
+  b.setInsertionPointToEnd(&body.front());
+  b.create<linalg::YieldOp>(loc,
+                            transposeOp.getRegion().front().getArgument(0));
+  return transposeOp;
+}
+
+LogicalResult readFromPackedValue(
+    OpBuilder& b, Location loc, Value packedValue, ShapedType targetTy,
+    ArrayRef<OpFoldResult> offsets, ArrayRef<int64_t> tileLevelsVec,
+    ArrayRef<int64_t> tileSizesVec, ArrayRef<int64_t> permutationVec,
+    InnerMostDimsInfo& info, Operation*& resultOp) {
+  auto packedTy = packedValue.getType().cast<ShapedType>();
+  int packedRank = packedTy.getRank();
+
+  auto const2IndexAttr = [&](int64_t val) {
+    return IntegerAttr::get(b.getIndexType(), val);
+  };
+
+  SmallVector<OpFoldResult> newOffsets(packedRank);
+  SmallVector<OpFoldResult> newSizes(packedRank);
+  SmallVector<OpFoldResult> newStrides(packedRank,
+                                       OpFoldResult{const2IndexAttr(1)});
+
+  AffineExpr s0, s1;
+  bindSymbols(b.getContext(), s0, s1);
+  AffineExpr floorDivExpr = s0.floorDiv(s1);
+  AffineExpr modExpr = s0 % s1;
+
+  int tileSizeIdx = 0;
+  int resultDimIdx = 0;
+  for (int dimIdx = 0; dimIdx < offsets.size(); ++dimIdx) {
+    int64_t level = tileLevelsVec[dimIdx];
+    OpFoldResult offset = offsets[dimIdx];
+    for (int localResultDimIdx = 0; localResultDimIdx <= level;
+         ++localResultDimIdx) {
+      int d = resultDimIdx + localResultDimIdx;
+      if (info.isInnerMostDim(d)) {
+        newOffsets[d] = const2IndexAttr(0);
+        newSizes[d] = const2IndexAttr(info.dimSizes[dimIdx]);
+      } else {
+        newSizes[d] = const2IndexAttr(1);
+        OpFoldResult tileSize =
+            const2IndexAttr(tileSizesVec[tileSizeIdx + localResultDimIdx]);
+        newOffsets[d] = makeComposedFoldedAffineApply(b, loc, floorDivExpr,
+                                                      {offset, tileSize});
+        offset =
+            makeComposedFoldedAffineApply(b, loc, modExpr, {offset, tileSize});
+      }
+    }
+    tileSizeIdx += level;
+    resultDimIdx += 1 + level;
+  }
+
+  if (!permutationVec.empty()) {
+    newOffsets =
+        disc_linalg_ext::interchange<OpFoldResult>(newOffsets, permutationVec);
+    newSizes =
+        disc_linalg_ext::interchange<OpFoldResult>(newSizes, permutationVec);
+  }
+
+  auto newShape = info.dimSizes;
+  newShape = disc_linalg_ext::interchange<int64_t>(newShape, info.permutation);
```

Comment:
why only use inner most tiled-dims to do interchange to make new shape?

Reply:
here we have already sliced a tile from the packed weight, thus only needs to transpose the tile (inner most dims) to replace the result of the original tile slice op.